pkg/meta/sql.go

```diff
@@ -4446,3 +4446,94 @@ func (m *dbMeta) loadDumpedACLs(ctx Context) error {
 		return nil
 	})
 }
+
+type dbDirHandler struct {
+	dirHandler
+}
+
+func (h *dbDirHandler) Insert(inode Ino, name string, attr *Attr) {
+	h.Lock()
+	defer h.Unlock()
+	if h.batch == nil {
+		return
+	}
+	if h.batch.isEnd || bytes.Compare([]byte(name), h.batch.maxName) < 0 {
+		h.batch.entries = append(h.batch.entries, &Entry{Inode: inode, Name: []byte(name), Attr: attr})
+		h.batch.indexes[name] = len(h.batch.entries) - 1
+	}
+}
+
+func (h *dbDirHandler) Delete(name string) {
+	h.Lock()
+	defer h.Unlock()
+
+	h.dirHandler.delete(name)
+	if h.batch != nil && !h.batch.isEnd && bytes.Compare(h.batch.maxName, []byte(name)) > 0 && h.batch.cursor != nil {
+		h.batch.cursor = h.batch.cursor.(int) - 1
+	}
+}
+
+func (m *dbMeta) newDirHandler(inode Ino, plus bool, entries []*Entry) DirHandler {
+	h := &dbDirHandler{
+		dirHandler: dirHandler{
+			inode:       inode,
+			plus:        plus,
+			initEntries: entries,
+			fetcher:     m.getDirFetcher(),
+			batchNum:    DirBatchNum["db"],
+		},
+	}
+	h.batch, _ = h.fetch(Background, 0)
+	return h
+}
+
+func (m *dbMeta) getDirFetcher() dirFetcher {
+	return func(ctx Context, inode Ino, cursor interface{}, offset, limit int, plus bool) (interface{}, []*Entry, error) {
+		entries := make([]*Entry, 0, limit)
+		err := m.roTxn(func(s *xorm.Session) error {
+			iCursor := offset
+			if cursor != nil {
+				iCursor = cursor.(int)
+			}
+
+			var ids []int64
+			if err := s.Table(&edge{}).Cols("id").Where("parent = ?", inode).Limit(limit, iCursor).Find(&ids); err != nil {
+				return err
+			}
+
+			s = s.Table(&edge{}).In("jfs_edge.id", ids).OrderBy("jfs_edge.name")
```

Comment:
These two queries for `edge` table can probably be merged as one.

Reply:
And ordered by `id` should be a simpler way.

Reply:
Need to sort by name, otherwise in some cases such as reverse order, the results may be incorrect

Reply:
The SQL query is splited which allows it to directly hit the index tree and reduce the amount of querying. 
Tried to merge statements in xorm, which would lose optimization effects.

A new pull request can be proposed for optimization.